### PR TITLE
Refactor POIs so that Mobs can be displayed

### PR DESCRIPTION
**Pull request by MajorBreakfast**
~~WIP~~

### CODE REVIEW

**Comment by MajorBreakfast** [comment]
mob-card

Note: This PR is a lot more than this simple layout. I've refactored the layout of the sightings card and a lot of related stuff.

You may review it now! :) Let's get it merged

**Comment by gyachdav** [comment]
Mucho cool!

**Comment by WoH** [comment]
There's 2 or more things going on in one PR again ...

**Comment by MajorBreakfast** [comment]
> more things going on in one PR

@WoH There's not. I never touched anything that wasn't related. If you want, you can think about this PR as "Refactor POIs so that Mobs can be displayed". If I had left things as they were, it would have been a mess. Nevertheless, plz review. I'd like to get this in. ;)

**Comment by WoH** [comment]
@MajorBreakfast I would've liked to have the initial refractoring in a PR.

**Comment by WoH** [comment]
I'd want @johartl to comment, the POI Cards are his baby.

**Comment by MajorBreakfast** [comment]
>  I would've liked to have the initial refractoring in a PR.

@WoH I've changed the name. The actual Mob card is the least part of this PR. However, there is little value in separating it out. I agree that @johartl should comment.

**Comment by MajorBreakfast** [comment]
I'm going to merge this this evening unless there are objections to any of the things I changed. The code is clean and I tested all affected app parts thoroughly, so it should be fine.

File: ionic2/app/components/map/map.component.ts
Comment (johartl) [inline ionic2/app/components/map/map.component.ts]: What's the reason for returning an empty mob object here?

File: ionic2/app/components/poi-card/poi-card.component.html
Comment (johartl) [inline ionic2/app/components/poi-card/poi-card.component.html]: By the definition of a mob there should always be multiple tweets so I think you can replace `tweet(s)` by `tweets`

File: ionic2/app/pages/map/map.page.html
Comment (johartl) [inline ionic2/app/pages/map/map.page.html:14]: I'm glad you added this button so I can test your functionality. But obviously that shouldn't end up in the final product. So I think either the button should be removed at all or somehow hidden such that it can only by used by developers.

**Review by johartl — CHANGES_REQUESTED** [review]

**Comment by johartl** [comment]
Regarding the discussion about too many things in one PR. I mostly agree that all changes focus around the mob card with the exception of renaming PokeSighting, PokeAttack, PokeGender, PokeMob, and some more. Personally, I think you could have done this in a separate PR.

File: ionic2/app/pages/map/map.page.html
Comment (MajorBreakfast) [reply ionic2/app/pages/map/map.page.html:14]: Let's leave it in for now, so that people can try it out. It's easy enough to remove later. It's for developers only, but triggering it through the console would be two cumbersome.

File: ionic2/app/components/map/map.component.ts
Comment (MajorBreakfast) [reply ionic2/app/components/map/map.component.ts]: Does the map already produce mobs? If so, what's the exact format of the event data?

File: ionic2/app/components/poi-card/poi-card.component.html
Comment (MajorBreakfast) [reply ionic2/app/components/poi-card/poi-card.component.html]: Okay. I'll change it.

File: ionic2/app/components/map/map.component.ts
Comment (johartl) [reply ionic2/app/components/map/map.component.ts]: I'm not sure if they already do. But the data should look like this https://github.com/PokemonGoers/PokeMap-2/issues/32#issuecomment-246179575 which is exactly what our `Mob` class represents. 
The best thing you can probably do is `const mob = Mob.fromObject(rawData);`

Comment (johartl) [reply ionic2/app/components/map/map.component.ts]: Well I just realized that you've removed the `fromObject` method.

File: ionic2/app/pages/map/map.page.html
Comment (WoH) [reply ionic2/app/pages/map/map.page.html:14]: Just make sure neither mock data nor the button remain after all.
I like being able to test it that way though.
This would be beautiful to test with a mock websocket (I believe that's where we get those).

File: ionic2/app/components/map/map.component.ts
Comment (WoH) [inline ionic2/app/components/map/map.component.ts]: no reason to use the function keyword

File: ionic2/app/pages/map/map.page.html
Comment (johartl) [reply ionic2/app/pages/map/map.page.html:14]: How is triggering it through the console cumbersome? And even if it is, this functionality will only ever be used for testing purposes by developers. So it doesn't have to be easily accessible by everyone.

File: ionic2/app/components/poi-card/poi-card.component.html
Comment (WoH) [inline ionic2/app/components/poi-card/poi-card.component.html]: you could make use of the Elvis operator here: `<div *ngIf="poi?.type == 'sighting' && pokemon">`

File: ionic2/app/components/poi-card/poi-card.component.html
Comment (WoH) [inline ionic2/app/components/poi-card/poi-card.component.html]: See above

File: ionic2/app/components/map/map.component.ts
Comment (MajorBreakfast) [reply ionic2/app/components/map/map.component.ts]: I removed it from the model itself because it doesn't really belong in there.

File: ionic2/app/components/map/map.component.ts
Comment (MajorBreakfast) [reply ionic2/app/components/map/map.component.ts]: Is that TypeScript feature?

File: ionic2/app/components/poi-card/poi-card.component.html
Comment (MajorBreakfast) [reply ionic2/app/components/poi-card/poi-card.component.html]: Elvis operator :) I'll change that

File: ionic2/app/components/poi-card/poi-card.component.html
Comment (MajorBreakfast) [reply ionic2/app/components/poi-card/poi-card.component.html]: Jep

File: ionic2/app/pages/map/map.page.html
Comment (MajorBreakfast) [reply ionic2/app/pages/map/map.page.html:14]: @WoH What do you think?

File: ionic2/app/components/map/map.component.ts
Comment (WoH) [reply ionic2/app/components/map/map.component.ts]: Yes, and we are using it all over the place I think :laughing: http://www.typescriptlang.org/docs/handbook/classes.html

File: ionic2/app/components/poi-card/poi-card.component.html
Comment (WoH) [reply ionic2/app/components/poi-card/poi-card.component.html]: I know that's something very Angular2 specific, but I think it is nice. Easy to understand, and makes these checks really concise.

File: ionic2/app/components/map/map.component.ts
Comment (MajorBreakfast) [reply ionic2/app/components/map/map.component.ts]: Compile error. Am I doing something wrong?

Comment (WoH) [reply ionic2/app/components/map/map.component.ts]: Can you paste the error?

Comment (MajorBreakfast) [reply ionic2/app/components/map/map.component.ts]: Leads to a boatload of errors. That's a bit long to post. However note that I made this a utility function instead of a static method.

Comment (MajorBreakfast) [reply ionic2/app/components/map/map.component.ts]: https://paste2.org/nW2f8j31

Comment (MajorBreakfast) [reply ionic2/app/components/map/map.component.ts]: TypeScript docu uses the keyword as well https://www.typescriptlang.org/docs/handbook/functions.html

File: ionic2/app/models/mob.ts
Comment (johartl) [inline ionic2/app/models/mob.ts:11]: If we extract this class to a separate file and call it `Tweet` we can reuse it in other parts of the application. For instance, the sentiment analysis also returns an array of tweets and these are not related to mobs.

File: ionic2/app/components/map/map.component.ts
Comment (WoH) [reply ionic2/app/components/map/map.component.ts]: The class ends before the function definition, I didn't realiize this on github.

Comment (MajorBreakfast) [reply ionic2/app/components/map/map.component.ts]: Yep, exactly.

File: ionic2/app/pages/map/map.page.html
Comment (MajorBreakfast) [reply ionic2/app/pages/map/map.page.html:14]: - I can remove it immediately, e.g. by uncommenting it
- Or keep it for a while, e.g. remove it before we launch/map shows mobs

**Comment by MajorBreakfast** [comment]
@johartl Plz update your review.

File: ionic2/app/pages/map/map.page.html
Comment (WoH) [reply ionic2/app/pages/map/map.page.html:14]: Maybe show it only in dev Mode?

File: ionic2/app/components/map/map.component.ts
Comment (MajorBreakfast) [reply ionic2/app/components/map/map.component.ts]: @johartl Should I make this a static method or a utility function? What do you prefer? I and @WoH need another opinion

Comment (johartl) [reply ionic2/app/components/map/map.component.ts]: I prefer the static method

File: ionic2/app/pages/map/map.page.html
Comment (johartl) [reply ionic2/app/pages/map/map.page.html:14]: Maybe set the style to `display: none` and then anyone can still look for the element in the DOM using the developer console and make it visible.

Comment (MajorBreakfast) [reply ionic2/app/pages/map/map.page.html:14]: Now it is only shown in dev mode

**Comment by MajorBreakfast** [comment]
Clear to merge? Any other wishes?

File: ionic2/app/services/config.service.ts
Comment (WoH) [inline ionic2/app/services/config.service.ts]:
```
  get buildEnvironment(): string {
    return env.BUILD_ENV;
  }
```

For consistency reasons this is better here maybe.

**Review by WoH — APPROVED** [review]

File: ionic2/app/services/config.service.ts
Comment (johartl) [reply ionic2/app/services/config.service.ts]: I agree. Even better: a method `isDevelopEnvironment()`

```
get isDevelopEnvironment(): boolean {
    return env.BUILD_ENV === 'develop';
}
```

But that's just my personal preference.

Comment (MajorBreakfast) [reply ionic2/app/services/config.service.ts]: Done

**Review by johartl — APPROVED** [review]

**Pull request merged**
